python/mxnet/gluon/block.py

```diff
@@ -1437,6 +1460,23 @@ def hybrid_forward(self, F, x, *args, **kwargs):
         # pylint: disable= invalid-name
         raise NotImplementedError
 
+    def reset_ctx(self, ctx):
+        """Re-assign all Parameters to other contexts. If the Block is hybridized, it will reset the _cached_op_args.
+
+        Parameters
+        ----------
+        ctx : Context or list of Context, default :py:meth:`context.current_context()`.
+            Assign Parameter to given context. If ctx is a list of Context, a
+            copy will be made for each context.
+        """
+        params = self.collect_params()
+        if self._cached_op:
+            for p in self._cached_op_args:
+                # resetting parameters creating by the partitioning backend
+                if p.name not in params:
+                    p.reset_ctx(ctx)
+        for p in params.values():
```

Comment:
Can you explain why we need to loop over `_cached_op_args` and only reset the params not in `params`, and then loop again over `params` and then reset them there instead? Is it possible to do the work in the 2nd loop in the first loop?

Reply:
Although i guess if we delete a param, then it will still be in `params` but not in `_cached_op_args`. And the context check will fail if we dont do all the params, so i guess this makes sense. Maybe we should have a comment that `params` and `_cached_op_args` might contain unique params (ie. there is no superset)?

Reply:
> Although i guess if we delete a param, then it will still be in params but not in _cached_op_args. 
That is the reason why
We don't want to do any additional reset_ctx: reset is costly because it copies NDArrays 

I will add some comments to clarify